h/services/group_members.py

```diff
@@ -60,18 +61,25 @@ def member_join(self, group, userid):
         if user in group.members:
             return
 
-        group.members.append(user)
+        group.memberships.append(GroupMembership(user=user))
 
         self.publish("group-join", group.pubid, userid)
 
     def member_leave(self, group, userid):
         """Remove `userid` from the member list of `group`."""
         user = self.user_fetcher(userid)
 
-        if user not in group.members:
+        matching_memberships = [
+            membership for membership in group.memberships if membership.user == user
+        ]
+
+        if not matching_memberships:
             return
 
-        group.members.remove(user)
+        for membership in matching_memberships:
+            self.db.delete(membership)
+            group.memberships.remove(membership)
+            user.memberships.remove(membership)
```

Comment:
If you just do `group.memberships.remove(membership)` that sets the `GroupMembership`'s `group_id` to `None` but doesn't remove it from the DB session, and then you end up with a `psycopg2.errors.NotNullViolation` on the `user_group.group_id` column. So it's necessary to also call `self.db.delete(membership)`.

The `user.memberships.remove(membership)` is _also_ needed because removing the membership from the group's memberships doesn't seem to automatically remove it from the user's memberships and you end up with some functional tests crashing when some code later in the request cycle is reading `user.memberships` and trying to read `user.memberships[i].group.name` and crashing because one of the memberships in `user.memberships` has `group=None`.